apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.YCSB/Program.cs

```diff
@@ -0,0 +1,513 @@
+                    {
+                        operation++;
+                        if (operation < s_operations.Count)
+                        {
+                            operationTarget -= s_operations[operation].Proportion;
+                        }
+                    }
+                    try
+                    {
+                        all.Add(s_operations[Math.Min(operation, s_operations.Count - 1)].Action(
+                            s_operations[Math.Min(operation, s_operations.Count - 1)].Latencies));
+                    }
+                    catch (Exception e)
+                    {
+                        DebugMessage($"This error is being ignored and we are continuing the test: {e.Message}");
+                    }
+                }
+            }
+
+            var timeout = Task.Delay(TimeSpan.FromSeconds(10));
+            await Task.WhenAny(Task.WhenAll(all), timeout);
+            s_totaloperations = all.Count;
+        }
+
+        /// <summary>
+        /// Runs a single worker task
+        /// </summary>
+        /// <param name="numWorkers"></param>
+        private static async Task RunWithTargetWorkers(int numWorkers)
+        {
+            //yield to let Main continue through.
+            await Task.Yield();
+            var parallelWorkers = new List<Task>();
+            var operation = ChooseOperation();
+
+            while (Interlocked.Decrement(ref s_operationsRemaining) > 0)
+            {
+                parallelWorkers.Add(s_operations[Math.Min(operation, s_operations.Count - 1)].Action(
+                    s_operations[Math.Min(operation, s_operations.Count - 1)].Latencies));
+                if (s_operationsRemaining % 1000 == 0)
+                {
+                    DebugMessage($"{s_operationsRemaining} operations remaining.");
+                }
+                operation = ChooseOperation();
+                if (parallelWorkers.Count >= numWorkers)
+                {
+                    try
+                    {
+                        parallelWorkers.Remove(await Task.WhenAny(parallelWorkers));
+                    }
+                    catch (Exception e)
+                    {
+                        DebugMessage($"This error is being ignored and we are continuing the test: {e.Message}");
+                        parallelWorkers.Clear(); //start over.
+                    }
+                }
+            }
+        }
+
+        private static int ChooseOperation()
+        {
+            var operationTarget = s_rand.Value.NextDouble() * s_total;
+            var operation = 0;
+            operationTarget -= s_operations[operation].Proportion;
+            while (operationTarget > 0 && operation < s_operations.Count)
+            {
+                operation++;
+                if (operation < s_operations.Count)
+                {
+                    operationTarget -= s_operations[operation].Proportion;
+                }
+            }
+            return operation;
```

Comment:
Make this return `s_operations[Math.Min(operation, s_operations.Count - 1)]`? It looks like this should make the calling code simpler to follow.